CHAOS-5004 DynamoDB replication pause fault (#7256)
harness/developer-hub: docs/chaos-engineering/chaos-faults/aws/dynamo-db-replication-pause.md

---
id: dynamo-db-replication-pause
title: Dynamo DB replication pause
---

Dynamo DB replication pause fault pauses the data replication in DynamoDB tables over multiple locations for the chaos duration.

> - When chaos experiment is being executed, any changes to the DynamoDB table will not be replicated in different regions, thereby making the data in the DynamoDB inconsistent.

Reviewer: ```suggestion
- When the chaos experiment is being executed, any changes to the DynamoDB table will not be replicated in different regions, thereby making the data in the DynamoDB inconsistent.
```

- You can execute this fault on a DynamoDB table that is global, that is, there should be more than one replica of the table.

![Dynamo DB replication pause](./static/images/dynamo-db-replication-pause.png)

## Use cases
Dynamo DB replication pause determines the resilience of the application when data (in a database) that needs to be constantly updated is disrupted.

### Prerequisites
- Kubernetes >= 1.17
- DynamoDB should be up and running.

> - The derived table name should be a global table.

Reviewer: If these are requirements/prereqs, you might consider using "must" instead of "should be."

> - DynamoDB should be up and running.
> - The derived table name should be a global table.

Reviewer: ```suggestion
- DynamoDB must be up and running.
- The derived table name must be a global table.
```

> - The Kubernetes secret should have AWS access configuration (key) in the `CHAOS_NAMESPACE`. A sample secret file looks like:

Reviewer: ```suggestion
- The Kubernetes secret must have AWS access configuration (key) in the `CHAOS_NAMESPACE`. A sample secret file looks like this:
```

  ```yaml
  apiVersion: v1
  kind: Secret
  metadata:
    name: cloud-secret
  type: Opaque
  stringData:
    cloud_config.yml: |-
      # Add the cloud AWS credentials respectively
      [default]
      aws_access_key_id = XXXXXXXXXXXXXXXXXXX
      aws_secret_access_key = XXXXXXXXXXXXXXX
  ```

:::tip
HCE recommends that you use the same secret name, that is, `cloud-secret`. Otherwise, you will need to update the `AWS_SHARED_CREDENTIALS_FILE` environment variable in the fault template and you won't be able to use the default health check probes.
:::

Below is an example AWS policy to execute the fault.

```json

> 

Reviewer: Is the JSON here pending?

Author: Yes, waiting on the JSON policy.

    </tr>
    <tr>
      <td> AWS_SHARED_CREDENTIALS_FILE </td>
      <td> Path to the AWS secret credentials. </td>
      <td> Default: <code>/tmp/cloud_config.yml</code>. </td>
      </tr>
  </table>

### Table names

Name of the DynamoDB **global** table on which you apply chaos. You can provide multiple table names as comma-separated values. Tune it using the `TABLE_NAMES` environment variable.

The following YAML snippet illustrates the use of this environment variable:

[embedmd]:# (./static/manifests/dynamo-db-replication-pause/table-names.yaml yaml)
```yaml
kind: Workflow
apiVersion: argoproj.io/v1alpha1
metadata:
  name: fis-dynamodb-replication-pause
  namespace: hce
spec:
  - name: fis-dynamodb-replication-pause
      components:
        env:
          - name: TOTAL_CHAOS_DURATION
            value: "60"
          - name: TABLE_NAMES
            value: "name-1,name-2"
          - name: REGION
            value: "us-east"
          - name: AWS_SHARED_CREDENTIALS_FILE
            value: /tmp/cloud_config.yml
        secrets:
          - name: cloud-secret
            mountPath: /tmp/
```

### Count

> Number of iterations of the chaos experiment to execute. The DynamoDB replication is paused these many times. Tune it using the `COUNT` environment variable.

Reviewer: ```suggestion
The number of iterations of the chaos experiment to execute. The DynamoDB replication is paused this many times. Tune it using the `COUNT` environment variable.
```
